patches/usb-gadget-interval-hid.patch

```diff
@@ -0,0 +1,152 @@
+-				      */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_descriptor_header *hidg_hs_descriptors[] = {
+@@ -191,10 +180,7 @@ static struct usb_endpoint_descriptor hidg_fs_in_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_IN,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 10, /* FIXME: Add this field in the
+-				       * HID gadget configuration?
+-				       * (struct hidg_func_descriptor)
+-				       */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_endpoint_descriptor hidg_fs_out_ep_desc = {
+@@ -203,10 +189,7 @@ static struct usb_endpoint_descriptor hidg_fs_out_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_OUT,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 10, /* FIXME: Add this field in the
+-				       * HID gadget configuration?
+-				       * (struct hidg_func_descriptor)
+-				       */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_descriptor_header *hidg_fs_descriptors[] = {
+@@ -749,6 +732,7 @@ static int hidg_bind(struct usb_configuration *c, struct usb_function *f)
+ 	struct device		*device;
+ 	int			status;
+ 	dev_t			dev;
++	unsigned short interval_ms = cpu_to_le16(hidg->interval);
+ 
+ 	/* maybe allocate device-global string IDs, and patch descriptors */
+ 	us = usb_gstrings_attach(c->cdev, ct_func_strings,
+@@ -779,16 +763,22 @@ static int hidg_bind(struct usb_configuration *c, struct usb_function *f)
+ 	hidg_interface_desc.bInterfaceSubClass = hidg->bInterfaceSubClass;
+ 	hidg_interface_desc.bInterfaceProtocol = hidg->bInterfaceProtocol;
+ 	hidg->protocol = HID_REPORT_PROTOCOL;
++	hidg_ss_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_ss_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
+ 	hidg_ss_in_comp_desc.wBytesPerInterval =
+ 				cpu_to_le16(hidg->report_length);
+ 	hidg_hs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_hs_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_fs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_fs_in_ep_desc.bInterval = interval_ms;
+ 	hidg_ss_out_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
+ 	hidg_ss_out_comp_desc.wBytesPerInterval =
+ 				cpu_to_le16(hidg->report_length);
++	hidg_ss_out_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_hs_out_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_hs_out_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_fs_out_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_fs_out_ep_desc.bInterval = interval_ms;
+ 	/*
+ 	 * We can use hidg_desc struct here but we should not relay
+ 	 * that its content won't change after returning from this function.
+@@ -924,6 +914,7 @@ CONFIGFS_ATTR(f_hid_opts_, name)
+ F_HID_OPT(subclass, 8, 255);
+ F_HID_OPT(protocol, 8, 255);
+ F_HID_OPT(report_length, 16, 65535);
++F_HID_OPT(interval, 16, 65535);
```

Comment:
why 16?

Reply:
fair point, up to 255ms should be enough :-)

Reply:
In case we want more than 255ms :-)

Reply:
yes :-)

Reply:
fixed